### DISCUSSION TITLE
Autoloading without Rails?

### DISCUSSION THREAD

**Discussion by jacob-carlborg** [rspec/rspec-rails#2762] [Q&A] [answered]
When installing rspec-rails, adding the gem and running `rails generate rspec:install`. rspec-rails will create some files, among them `spec_helper.rb` and `rails_helper.rb`. As far as I understand, the reason for having two separate files is to be able to avoid booting the whole Rails applications when testing parts of the application that does not depend on Rails. Then you'll only require the `spec_helper` file in those tests that test things that don't depends on Rails.

My question is then, does anyone use this? How are you dealing with requiring files? Usually in a Rails application you don't need to use `require` because Rails autoloads your files. But when Rails is not booted there's nothing that autoloads the files. Are you adding `require` to your files? Ending up with some files needing `require` and some not. Do you manually configure Zeitwerk in `spec_helper` to allow to continue using the autoloader. How does that then work when requiring `rails_helper` that will boot the Rails application and also configure Zeitwerk?

**Comment by JonRowe** [comment]
If you run the whole suite you will end up requiring rails_helper and everything is "as normal", if you run an individual spec file which just requires spec_helper and you haven't setup Ruby's own autoload, typically I'd recommend you require just the files you need in the spec

**Reply by jacob-carlborg** [reply]
> if you run an individual spec file which just requires spec_helper and you haven't setup Ruby's own autoload

How would Ruby's own autoload interact with Zeitwerk?

> typically I'd recommend you require just the files you need in the spec

It's not just in the specs, it's in the files being tested as well. Example, two classes: `A` and `B`, each in their own files. `A` depends on `B`. Sure I can add a `require "b"` in `a.rb`. Then there's the problem in Ruby in which order files are required. In a gem that is usually solved by requiring all files in the file that is the entrypoint for the gem and not require in any other files. But there's no natural entrypoint for a Rails application in the same way. Unless you count `spec_helper.rb`.

**Reply by pirj** [reply]
Are you asking in general, or have some specific example in mind?
I suggest you harvest repos listed in real-world-rspec and see those.
At a glance, discourse and gitlab prefer to load everything. Bootsnap and faster computers might have reduced the urge to separate. And I haven’t seen many people use granular bundler groups or be careful about requiring gems they directly use at the top of files. It’s not Python, we have no runtime isolation.

**Reply by jacob-carlborg** [reply]
GitLab seems to do its own thing. It has a `spec_helper.rb` but no `rails_helper.rb`. But it does have a a file called [`fast_spec_helper.rb`](https://gitlab.com/gitlab-org/gitlab/-/blob/master/spec/fast_spec_helper.rb). This file requires [`rails_autoload`](https://gitlab.com/gitlab-org/gitlab/-/blob/master/spec/rails_autoload.rb) which uses Zeitwerk to autoload some paths. So GitLab's `fast_spec_helper.rb` seem to correspond to the conventional `spec_helper.rb` and GitLab's `spec_helper.rb` seems to correspond to the conventional `rails_helper.rb`. 

Discourse seems to only be using `rails_helper.rb` and no `spec_helper.rb`.

I probably going autoload using Zeitwerk as GitLab is doing.

**Reply by pirj** [reply]
Does this answer your initial question?

**Reply by JonRowe** [reply]
> How would Ruby's own autoload interact with Zeitwerk?

AFAIK it just works, its much more declarative than Rails autoloaders. Its pretty common for gems to use it.

> It's not just in the specs, it's in the files being tested as well. Example, two classes: A and B, each in their own files. A depends on B. Sure I can add a require "b" in a.rb. 

You can solve this in the files, but given in reality you will be relying on zeitwerk its best to just require b, then a, in the spec for A. Theres no harm in repitition across spec files because requiring an already required file is a no-op.


> Then there's the problem in Ruby in which order files are required. In a gem that is usually solved by requiring all files in the file that is the entrypoint for the gem and not require in any other files.

Thats not at all true, we use autoload for example to require files as needed and have multiple splits due to how various extensions / dependencies work. Basically any file that depends on other files will trigger a require of that depended file which again is a no-op if already required.

> But there's no natural entrypoint for a Rails application in the same way. Unless you count spec_helper.rb.

Depending on what you mean, if you mean booting the application there is in the form of `config/environment.rb` the same way that Rails is bootstrapped when running in a rackup file or other boot, if you mean for zeitwerk et al autoloading there is not and you shouldn't be using spec_helper for those scenarios..

Loading spec helper only is for things where you control the load sequence

**Reply by jacob-carlborg** [reply]
> Does this answer your initial question?

No, I don't think I have got a clear answer.

**Reply by jacob-carlborg** [reply]
> You can solve this in the files, but given in reality you will be relying on zeitwerk its best to just require b, then a, in the spec for A

Seems like a bad idea. Then the test would start relying on implementation details.

> Thats not at all true, we use autoload for example to require

If you use an autoloader, there's no problem. But if you don't, there's definitely a problem.

> Depending on what you mean, if you mean booting the application there is in the form of config/environment.rb

I mean when not booting the Rails application. That's what this whole discussion is about.

> if you mean for zeitwerk et al autoloading there is not and you shouldn't be using spec_helper for those scenarios..
Loading spec helper only is for things where you control the load sequence

That's disappointing 😞.

**Reply by jacob-carlborg** [reply]
I tried to set up Zeitwerk in `spec_helper` and run into a problem. When running tests in multiple files and some require `spec_helper` and some require `rails_helper`. There can only be one autoloader for a given path, Zeitwerk raises an exception if two autoloaders (one from `spec_helper` and one from Rails`) tries to manage the same path. I did manage to find a workaround, but it feels ugly.

This whole idea of splitting up the files does not seem to work out that well in practice.

**Reply by JonRowe** [reply, marked as answer]
Thats why I said you need to not use split helpers for Zeitwerk, and the same holds true for most gems that integrate tightly with Rails. 

Splitting the spec helpers is for people who are comfortable without e.g. zeitwerk, if you are working with "PORO" (plain old ruby object) codebases within Rails apps it offers an easy way to run plain rspec tests without Rails.

**Reply by jacob-carlborg** [reply]
I guess the last reply answers my question.
